parsons/dbt/dbt.py

```diff
@@ -0,0 +1,268 @@
+
+    def __init__(
+        self,
+        slack_channel: Optional[str] = None,
+        slack_webhook: Optional[str] = None,
+        slack_api_key: Optional[str] = None,
+    ):
+        self.start = time.time()
+        self.log_messages = []
+        self.error_messages = []
+        self.done_messages = []
+        self.slack_channel = slack_channel
+        self.slack_webhook = slack_webhook
+        self.slack_api_key = slack_api_key
+
+    def record_start(self, command: str) -> None:
+        """Record start time for command"""
+        self._command_times[command] = {"start": time.time()}
+
+    def record_end(self, command: str) -> None:
+        """Record end time for command"""
+        self._command_times[command]["end"] = time.time()
+
+    def seconds_to_time_string(self, seconds: int):
+        time_str = ""
+        command_time = time.gmtime(seconds)
+        if command_time.tm_yday - 1:
+            time_str += f"{command_time.tm_yday - 1} days, "
+        if command_time.tm_hour:
+            time_str += f"{command_time.tm_hour} hours, "
+        if command_time.tm_min:
+            time_str += f"{command_time.tm_min} minutes, "
+        if command_time.tm_sec:
+            time_str += f"{command_time.tm_sec} seconds"
+
+        return time_str
+
+    def record_result(self, command: str, error_messages: list[str], done_message: str):
+        command_seconds = int(
+            self._command_times[command]["end"] - self._command_times[command]["start"]
+        )
+
+        log_message = ""
+        if error_messages:
+            log_message += ":red_circle:"
+            status = "Error"
+        else:
+            log_message += ":large_green_circle:"
+            status = "Success"
+
+        time_str = self.seconds_to_time_string(command_seconds)
+        log_message += f"Invoke dbt with `dbt {command}` ({status} in {time_str})"
+        if done_message:
+            log_message += f"\n*Summary*: `{done_message}`"
+
+        if error_messages:
+            log_message += "\nError messages:\n```{}```".format(
+                "\n".join(error_messages)
+            )
+
+        self.log_messages.append(log_message)
+
+    def send_to_slack(self) -> None:
+        """Log final result to logger and send to slack."""
+        end_time = time.time()
+        duration_seconds = int(end_time - self.start)
+        duration_time_str = self.seconds_to_time_string(duration_seconds)
+
+        full_log_message = ""
+        if any(":red_circle:" in log_message for log_message in self.log_messages):
+            status = "failed"
+            full_log_message += ":red_circle:"
+        else:
+            status = "succeeded"
+            full_log_message += ":large_green_circle:"
+
+        full_log_message += f"*dbt run {status}*"
+        full_log_message += f"\n*Duration:* {duration_time_str}\n\n"
+        full_log_message += "\n".join(self.log_messages)
+
+        if self.slack_webhook:
```

Comment:
I think this is the main (only?) place where the logger is calling Slack directly, so it should be straightforward to lift out. I think the main challenge is handling the conditional here, on whether to use the webhook or the API key.

Also `:red_circle:` etc is going to be platform-specific but I don't think it's a big deal if a notification has `:red_circle:` instead of an actual red circle.

Reply:
As written, if no slack configs are provided (no API key or webhook is passed in), this if/else block will not execute on any code and logging will remain limited to stdout.

Parsons has a few other notifications interfaces, but for a few reasons I don't think it's important to generalize the notifications of this particular utility over them at this time:
- We discussed this PR in the TMC engineering working group, and no one had a use case besides Slack for notifications
- Parsons itself doesn't have a generic notifications API, making it a bit cumbersome to try and build one into this PR